src/JsonSchema/Uri/UriResolver.php

```diff
@@ -81,77 +34,124 @@ public function resolve($uri, $baseUri = null)
             return $baseUri;
         }
 
-        $components = $this->parse($uri);
-        $path = $components['path'];
-        
-        if (! empty($components['scheme'])) {
+        $parts = $this->parse($uri);
+
+        // Return if uri is already absolute
+        if (isset($parts['scheme'])) {
             return $uri;
         }
-        $baseComponents = $this->parse($baseUri);
-        $basePath = $baseComponents['path'];
-        
-        $baseComponents['path'] = self::combineRelativePathWithBasePath($path, $basePath);
-        if (isset($components['fragment'])) {
-            $baseComponents['fragment'] = $components['fragment'];
+
+        $uri = $this->parse($baseUri);
+
+        // Replace scheme and host if specified in the new uri
+        if (isset($parts['scheme'])) {
+            $uri['scheme'] = $parts['scheme'];
+        }
+        if (isset($parts['host'])) {
+            $uri['host'] = $parts['host'];
         }
 
-        return $this->generate($baseComponents);
-    }
-    
-    /**
-     * Tries to glue a relative path onto an absolute one
-     *
-     * @param string $relativePath
-     * @param string $basePath
-     * @return string Merged path
-     * @throws UriResolverException
-     */
-    public static function combineRelativePathWithBasePath($relativePath, $basePath)
-    {
-        $relativePath = self::normalizePath($relativePath);
-        if ($relativePath == '') {
-            return $basePath;
+        // Join the base URI path with the new path
+        if (isset($parts['path'])) {
+            if (isset($uri['path'])) {
+                if (substr($uri['path'], -1) !== '/') {
+                    $uri['path'] = rtrim(str_replace(basename($uri['path']), '', $uri['path']), '/') . '/';
+                }
+                $uri['path'] .= ltrim($parts['path'], '/');
+            } else {
+                $uri['path'] = $parts['path'];
+            }
         }
-        if ($relativePath{0} == '/') {
-            return $relativePath;
+
+        // Replace query and fragments
+        if (isset($parts['query'])) {
+            $uri['query'] = $parts['query'];
+        }
+        if (isset($parts['fragment'])) {
+            $uri['fragment'] = $parts['fragment'];
         }
 
-        $basePathSegments = explode('/', $basePath);
+        return $this->build($uri);
+    }
+
+    public function extractLocation($uri)
```

Comment:
What do you think about moving this and `extractFragment()` to a private method of `RefResolver`? I don't see much reusability here.